### PR TITLE
Bump the minimal OCaml version to 4.09.0.

### LINKED ISSUE #14260: [ocaml] Version bump for 8.16 ?
Hi folks,

there was some discussion about bumping the supported version of OCaml for 8.15 [IMHO 8.14 is out of the question]

I am opening this issue to centralize discussion.

There are several options here, the way I can see it:
- stay at 4.05: *pros:* stability, *cons*: nothing that is a blocker IMO , so we could stay
- go to 4.07: *pros*: `Stdlib` module, some issues solved, better Stdlib, , *cons* still a bit old
- 4.08: *pros*: monadic syntax, even more improvements for stdlib; *cons* bug in VM setting parameters [fixed in > 4.11.2], slowdown https://github.com/coq/coq/issues/11652 (fixed in >=8.12)
- 4.10: *cons* https://github.com/ocaml/ocaml/issues/9606 slow ocamldebug (also 4.11, 4.12)
- 4.11: *pros*: good baseline for the next 3-4 years; many bugfixes and improvements, a couple of regressions *cons* too new?
- 4.12: *cons*: https://github.com/ocaml/ocaml/issues/10517 ocamldebug doesn't work (also 4.13)

Regarding distro support, stable distros should ship with 4.11 for the 8.15 timeline, IIANM.

### PR DESCRIPTION
**Pull request by ppedrot**
As decided in common.

Fixes #14260.

### CODE REVIEW

**Comment by ppedrot** [comment]
The unimath failure is spurious.

**Comment by ppedrot** [comment]
There are other cleanups documented in the code that can benefit of this bump, but I don't think that it makes sense to make it part of this PR. Hence, I believe that this PR is ready and in need of an assignee.

**Comment by ppedrot** [comment]
Fiat-crypto needs an overlay, or alternatively we may need to add the ocaml-shims package to the docker image. What do you think is better?

**Comment by JasonGross** [comment]
> Fiat-crypto needs an overlay, or alternatively we may need to add the ocaml-shims package to the docker image. What do you think is better?

The warning is a red herring, the actual error is
```
ocamlopt.opt got signal and exited
Command exited with non-zero status 2
src/ExtractionOCaml/bedrock2_word_by_word_montgomery (real: 88.51, user: 63.03, sys: 3.06, mem: 2435932 ko)
```
This looks like an OOM, perhaps due to the performance regression you reported a while ago?  Maybe bump the OCaml version on the Fiat Crypto OCaml job to 4.13 and drop a link to the OCaml issue in a comment?

**Comment by ppedrot** [comment]
Incidentally, #15946 (which fixes the GC regression) compiles fiat-crypto-ocaml correctly.

**Comment by ppedrot** [comment]
Rerunning the CI was enough to make it pass. It might be near the OOM limit, but it's not always hitting it.

**Comment by JasonGross** [comment]
> Incidentally, #15946 (which fixes the GC regression) compiles fiat-crypto-ocaml correctly.

This is a red herring, there is no coqc process running when the OOM is hit (what the job is testing is that Coq extraction generates working code).  



> Rerunning the CI was enough to make it pass. It might be near the OOM limit, but it's not always hitting it.

Maybe you want to switch to -j1 for Fiat Crypto OCaml?  Otherwise you'll probably keep seeing periodic failures based on timing...

**Comment by ppedrot** [comment]
> Maybe you want to switch to -j1 for Fiat Crypto OCaml?

I suggest deferring this to another PR, and wait to see if this is really reproducible before trying to fix it.

**Comment by ppedrot** [comment]
This will conflict with #15948. If somebody is kind enough to merge this one first, that will be one less hurdle for the kind 8.16 RM...

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Comment by ppedrot** [comment]
Thanks!

File: dev/ci/docker/bionic_coq/Dockerfile [hunk 1]
```diff
@@ -39,15 +39,15 @@ ENV NJOBS="2" \
...
+ENV COMPILER="4.09.0"
 
 # Common OPAM packages
 ENV BASE_OPAM="zarith.1.12 ocamlfind.1.9.1 ounit2.2.2.3 odoc.1.5.3" \
     CI_OPAM="ocamlgraph.1.8.8 yojson.1.7.0 cppo.1.6.8" \
     BASE_ONLY_OPAM="dune.2.7.1 elpi.1.14.1 stdlib-shims.0.1.0"
 
 # BASE switch; CI_OPAM contains Coq's CI dependencies.
-ENV COQIDE_OPAM="cairo2.0.6.1 lablgtk3-sourceview3.3.1.0"
+ENV COQIDE_OPAM="cairo2.0.6.1 lablgtk3-sourceview3.3.1.2"
```
Comment (SkySkimmer) [inline dev/ci/docker/bionic_coq/Dockerfile:50]: Since this got bumped should we update install.md?

Comment (Zimmi48) [reply dev/ci/docker/bionic_coq/Dockerfile:50]: Yes, please!

Comment (Zimmi48) [reply dev/ci/docker/bionic_coq/Dockerfile:50]: And the changelog entry should be edited to include this information.

**Pull request merged**